### DISCUSSION TITLE
Running the server inside a container

### DISCUSSION THREAD

**Discussion by pocsuka** [dotnet/WatsonWebserver#52] [Q&A] [answered]
Hello,

I have an issue making calls to the server, if it runs inside of a docker container. I hope you might be able to help solving this issue.

In my project the docker img basically contains only the executables.

The Dockerfile is the following:
FROM debian:stable-slim
WORKDIR /app
ADD . .
CMD ["./my_exe"]
EXPOSE 9000

Whit this Dockerfile but with a different server library, I was able to make api calls to the server.

Simply just running the exe, I can make api calls to the server, but if it runs in a container I don't have any success.
If I am inside of the container, I am able to make curl calls to the server, and it works fine.

I have absolutely no idea what can be the issue.

Thanks in advance.

Best regards,
David

**Comment by jchristn** [comment, marked as answer]
Hi @pocsuka could you try using the ```dockerfile``` in the ```Test.Docker``` project and let me know the result?  https://github.com/jchristn/WatsonWebserver/tree/master/Test.Docker

**Reply by pocsuka** [reply]
Can't use that. My project it is a Unity3D app, so I have to first build the app, and when it's completed create a docker img from it.

**Reply by jchristn** [reply]
What does your constructor look like, i.e. on which hostname are you listening, and what port?  The dockerfile looks fine, though I'd suggest changing to ```EXPOSE 9000/tcp```

**Reply by pocsuka** [reply]
I have tried with 9000/tcp as well
I have created a wrapper class, which got a Server member
```javascript
.
.
Server server;
string hostname = "localhost";
int port = 9000;
.
.
public void start()
{
      server = new Server(hostname, port, false);
      //adding the routes
      server.start();
}
.
.
```

I have tried other ports as well

**Reply by jchristn** [reply]
Could you try listening on ```0.0.0.0``` (running with admin rights within the container) and see if that works?

**Reply by pocsuka** [reply]
If I am within the container I am able to make curl calls to the server, if it runs on localhost. On 0.0.0.0 I have received a "connection refused" msg after executing the curl call

**Reply by jchristn** [reply]
Got it.

Could you try using ```*``` as well as ```127.0.0.1``` instead of ```localhost```?  The HTTP HOST header has to match the listener specified.  ```*``` with root privileges should match anything (some OSes use ```*```, some use ```0.0.0.0```, some use either).

**Reply by pocsuka** [reply]
127.0.0.1 behaves same as localhost, can call the server from within the container, but not from outside, how can I call * exactly? just replace the host with that character? if that's the way it have sent back a "connection refused"

**Reply by jchristn** [reply]
Yes you would just supply ```*``` as a character.  Which OS and runtime are you using?

**Reply by pocsuka** [reply]
I have a windows machine, I also tried ubuntu terminal on windows, same results, if I ran the executable, it worked, but if it was in a container I was not able to make any calls to the api (sadly I don't have access to a linux machine) the Unity3d project is set to netstandard2.0, as far as I know it uses Mono in the back, the docker base img is debian:stable-slim

**Reply by jchristn** [reply]
Ok - thank you.  I'm spinning up a couple new VMs so I can revisit Docker.  Out of curiosity, what was the other server library that worked for you?  I'd like to understand if it uses http.sys under the hood.  On my end, I'll report back as soon as I can get the environment ready and tested.

**Reply by pocsuka** [reply]
I have checked, so the previous "lib" was not a lib at all, it was just a System.Net.HttpListener

**Reply by jchristn** [reply]
Cool - thank you.  Then it should work, since Watson uses exactly that: https://github.com/jchristn/WatsonWebserver/blob/master/WatsonWebserver/WatsonWebserver.cs

Will report back as soon as I get the environment set up.

**Reply by jchristn** [reply]
Quick update, I had it working on Ubuntu 14.04 but it's failing on Ubuntu 20.04.  I have a Windows installation set up to test it, will be working on that next.  Are you forced to use ```FROM debian:stable-slim```?

**Reply by pocsuka** [reply]
Thanks, cool.

I don't think so, or at least I hope I am not forced to use that.

I have tried with an Ubuntu 14.04 base img, but I was not able to make api calls to the server. Have you done any modifications on the server's source?

**Reply by jchristn** [reply]
No modifications to the source.  Will keep you posted!

**Reply by jchristn** [reply]
I think what's happening is the ```AcceptConnections``` task is throwing an exception.  I'll be debugging as I can today and tomorrow.  I added handlers for the ```ExceptionEncountered``` and ```ServerStopped``` events to my ```Test.Docker``` project, and I'm getting this:
```
C:\Code\watsonwebserver\Test.Docker\bin\Debug\netcoreapp3.1>test.docker
Watson Webserver started on http://0.0.0.0:8080
System.Net.HttpListenerException (50): The request is not supported.
   at System.Net.HttpListener.AddPrefixCore(String registeredPrefix)
   at System.Net.HttpListener.AddAllPrefixes()
   at System.Net.HttpListener.Start()
   at WatsonWebserver.Server.AcceptConnections(CancellationToken token) in C:\Code\watsonwebserver\WatsonWebserver\WatsonWebserver.cs:line 348
Server stopped
```
But at least I think I have a starting point.  This will likely require a code change and an upgrade to the library.  It could end up being something environmental with the VMs I spun up to do the test because running in Docker was not a problem previously.  I'll keep you posted.

**Reply by jchristn** [reply]
Quick update, I think I know what changed.  Everything is converted into a ```Uri``` now internally (this started when I added the ```WatsonWebserverSettings``` class a couple versions ago I believe).  The issue is that using a listener like ```*```, ```+```, or ```0.0.0.0``` will throw a ```System.UriFormatException```, and this happens within the ```AcceptConnections``` task when the getter is called on the URI list.  

Could you set the ```ExceptionEncountered``` and ```ServerStopped``` events and see if they fire, and if they do, what you see in the ```ExceptionEventArgs.Exception``` property (i.e. ```.ToString()``` it and paste here)?

**Reply by pocsuka** [reply]
How can I set events? What should I write and where?

**Reply by jchristn** [reply]
Here's how I did it (```_Server``` is my global for Watson):
```
_Server.Events.ExceptionEncountered += ExceptionEncountered;
_Server.Events.ServerStopped += ServerStopped;

static void ExceptionEncountered(object sender, ExceptionEventArgs args)
{
    Console.WriteLine(args.Exception.ToString());
}

static void ServerStopped(object sender, EventArgs args)
{
    Console.WriteLine("Server stopped");
}
```

**Reply by pocsuka** [reply]
I have added this to my code, but it does nothing(after calling to the server nothing have been logged). I assume I did not do it in a correct way.
Here is my code:
```cs
public class MyServer
{    
    public static Server server;    
    public void start()
    {
        server = new Server("127.0.0.1", 9000, false);        
        server.Events.ExceptionEncountered += ExceptionEncountered;
        server.Events.ServerStopped += ServerStopped;
        //adding routes
        logger.Info("server started at: " + server.Settings.Hostnames[0] + " port: " + server.Settings.Port);        
        server.Start();
    }
    
    static void ExceptionEncountered(object sender, ExceptionEventArgs args)
    {        
        logger.Info(args.Exception.ToString());
    }

    static void ServerStopped(object sender, EventArgs args)
    {        
        logger.Info("Server stopped");
    }
}
```

**Reply by jchristn** [reply]
That looks correct; I don't think you're doing it wrong. I'm working on a new version that won't use Uri objects anyway which should resolve the listener issue. Should have it ready before end of the weekend. Sorry this is taking so long.

**Reply by jchristn** [reply]
Hi @pocsuka I just pushed v4.1.0 to NuGet, would you mind giving it a try?  You'll want to set your listener to run on one of ```*```, ```+```, or ```0.0.0.0```, and will have to run it within the container with elevated privileges.  This, at minimum, fixes the URI format exception for listening on multiple hostnames, and may be the fix for the listener not starting.  Thanks!

**Reply by jchristn** [reply]
Ok, another quick update.

I tested with 4.1.0 using the ```Test.Docker``` project.  I had to remove the project reference back to the Watson source, and instead use the NuGet package.  Using the ```Dockerfile``` in the project source, I was able to successfully connect to the instance running in Docker, on Windows 10, using .NET Core 3.1.

```
C:\Code\watsonwebserver\Test.Docker>docker build -t watsontest -f Dockerfile .
[+] Building 24.6s (16/16) FINISHED
 => [internal] load build definition from Dockerfile                                                               0.1s
 => => transferring dockerfile: 32B                                                                                0.0s
 => [internal] load .dockerignore                                                                                  0.0s
 => => transferring context: 2B                                                                                    0.0s
 => [internal] load metadata for mcr.microsoft.com/dotnet/core/sdk:3.1                                             0.7s
 => [internal] load metadata for mcr.microsoft.com/dotnet/core/runtime:3.1                                         0.7s
 => [build 1/7] FROM mcr.microsoft.com/dotnet/core/sdk:3.1@sha256:f7bc4a555fef18101b64bc7b954738489349d53655731ef  0.0s
 => [internal] load build context                                                                                  0.2s
 => => transferring context: 1.18MB                                                                                0.2s
 => [stage-1 1/3] FROM mcr.microsoft.com/dotnet/core/runtime:3.1@sha256:535ca49e393bfed05eef1e1f0458eeda4057c12a1  0.0s
 => CACHED [stage-1 2/3] WORKDIR /app                                                                              0.0s
 => CACHED [build 2/7] WORKDIR /app                                                                                0.0s
 => [build 3/7] COPY *.csproj .                                                                                    0.1s
 => [build 4/7] RUN dotnet restore                                                                                 8.8s
 => [build 5/7] COPY . ./                                                                                          0.3s
 => [build 6/7] RUN dotnet build   -f netcoreapp3.1 -c Release                                                     9.9s
 => [build 7/7] RUN dotnet publish -f netcoreapp3.1 -c Release -o out                                              4.0s
 => [stage-1 3/3] COPY --from=build /app/out .                                                                     0.1s
 => exporting to image                                                                                             0.2s
 => => exporting layers                                                                                            0.2s
 => => writing image sha256:244cf6d3a5db238003ed44a5b3625fafd79649b34b343bd3d0965696cfdf12f9                       0.0s
 => => naming to docker.io/library/watsontest                                                                      0.0s

C:\Code\watsonwebserver\Test.Docker>docker run --user root -p 8080:8080 watsontest
Watson Webserver started on http://*:8080
```

![image](https://user-images.githubusercontent.com/1421081/107136997-fcd6ad80-68bc-11eb-9548-408043de19eb.png)

Dockerfile: https://github.com/jchristn/WatsonWebserver/blob/master/Test.Docker/Dockerfile

Note I was copying in source, building, publishing, etc.  You may not have to do these steps.

Please let me know if this works for you.  In my case, I was listening on ```*```.

**Reply by pocsuka** [reply]
Thanks!

I will check it in the afternoon (where I live currently 6am :D)
Could you clarify please what does "will have to run it within the container with elevated privileges" mean exactly? What should I do?

**Reply by jchristn** [reply]
Hi, understood, sorry to send it so early your time.  Since the listener will need to listen for requests on any hostname, you'll need to specify the listener hostname of *, +, or 0.0.0.0.  Doing so requires that the process run under elevated privileges (this is an OS limitation, not a Watson limitation).  

For my case, I had to use ```--user root``` in my ```docker run``` statement.

```
docker run --user root -p 8080:8080 watsontest
```

**Reply by pocsuka** [reply]
It's working!
I had to check it before I go to sleep :D
Big thanks, really

**Reply by jchristn** [reply]
That's awesome to hear!  I'm going to close this out - please reopen if any more issues :)

**Reply by jchristn** [reply]
Hi @pocsuka I would be remiss if I didn't tell you that running the process as root within the container introduces some security concerns.  You really shouldn't run it as root, it's just an easy way to get the HTTP server to "listen for anything" (i.e. *, +, 0.0.0.0, etc).

Ideally you'd specify an IP address or hostname on which it would listen, and make sure that the container routing forwarded requests to the webserver using that specific hostname.  One of the drawbacks of anything using Windows http.sys (like Watson) is that, unless you're using */+/0.0.0.0, the HTTP HOST header of the incoming request *MUST* match exactly the listener prefix.  

So, net-net, use root at your own risk.  I'd definitely recommend finding out what HOST header is supplied on incoming requests, and set your listener in Watson to use that instead of */+/0.0.0.0.

Cheers, Joel

**Reply by pocsuka** [reply]
Thanks for the information.

Our service will run in an internal network and we also authenticate every request, so I think running the container with root should be okay, but the DevOps guys might think differently :D
